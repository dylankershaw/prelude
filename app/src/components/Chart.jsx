import React, { useEffect, useRef } from 'react';
import Chart from 'chart.js';

export default ({ data }) => {
  const ref = useRef(null);

  useEffect(() => {
    const myChartRef = ref.current.getContext('2d');

    new Chart(myChartRef, {
      type: 'line',
      data: {
        labels: Object.keys(data).reverse(),
        datasets: [
          {
            label: 'Songs Played',
            data: Object.values(data).reverse()
          }
        ]
      },
      options: {
        //Customize chart options
      }
    });
  }, [data]);

  return (
    <div>
      <canvas id="myChart" ref={ref} />
    </div>
  );
};
