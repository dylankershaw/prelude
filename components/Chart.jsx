import React, {useEffect, useRef, useState} from 'react';
import Chart from 'chart.js';

export default ({songData, selectedArtist}) => {
  const [chartData, setChartData] = useState([]);
  const [fullPlaysOnly, setFullPlaysOnly] = useState(false);
  const ref = useRef(null);

  useEffect(() => {
    if (!selectedArtist) return setChartData([]);

    let songs = songData.filter(s => s.master_metadata_album_artist_name?.includes(selectedArtist));

    if (fullPlaysOnly) songs = songs.filter(({reason_end}) => reason_end === 'trackdone');

    const data = songs.reduce((acc, cur) => {
      const date = cur.ts.split(' ')[0];
      return {...acc, [date]: (acc[date] || 0) + 1};
    }, {});

    setChartData(data);
  }, [songData, selectedArtist, fullPlaysOnly]);

  useEffect(() => {
    const myChartRef = ref.current.getContext('2d');

    new Chart(myChartRef, {
      type: 'line',
      data: {
        labels: Object.keys(chartData).reverse(),
        datasets: [
          {
            label: 'Songs Played',
            data: Object.values(chartData).reverse()
          }
        ]
      },
      options: {
        //Customize chart options
      }
    });
  }, [chartData]);

  return (
    <div>
      <input
        id='full-plays-only'
        type='checkbox'
        checked={fullPlaysOnly}
        onChange={e => setFullPlaysOnly(e.target.checked)}
      />
      <label htmlFor='full-plays-only'>Only include songs played in full</label>
      <canvas id='myChart' ref={ref} />
    </div>
  );
};
