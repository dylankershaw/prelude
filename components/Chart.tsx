import React, {useEffect, useRef, useState} from 'react';
import Chart from 'chart.js';
import {Song, ChartData} from '../Types';

interface Props {
  songData: Array<Song>;
  selectedArtist: string;
}

const ChartComponent: React.FC<Props> = ({songData, selectedArtist}) => {
  const [chartData, setChartData] = useState<ChartData>({});
  const [fullPlaysOnly, setFullPlaysOnly] = useState(false);
  const ref = useRef<HTMLCanvasElement>(null);

  useEffect(() => {
    if (!selectedArtist) return setChartData({});

    let songs = songData.filter((s: Song) =>
      s.master_metadata_album_artist_name?.includes(selectedArtist)
    );

    if (fullPlaysOnly) songs = songs.filter(({reason_end}) => reason_end === 'trackdone');

    const data = songs.reduce((acc: ChartData, cur) => {
      const date = cur.ts.split(' ')[0];
      return {...acc, [date]: (acc[date] || 0) + 1};
    }, {});

    setChartData(data);
  }, [songData, selectedArtist, fullPlaysOnly]);

  useEffect(() => {
    const chartRef = ref?.current?.getContext('2d');
    if (!chartRef) return;

    new Chart(chartRef, {
      type: 'line',
      data: {
        labels: Object.keys(chartData).reverse(),
        datasets: [
          {
            label: 'Songs Played',
            data: Object.values(chartData).reverse()
          }
        ]
      },
      options: {
        //Customize chart options
      }
    });
  }, [chartData]);

  return (
    <div>
      <input
        id='full-plays-only'
        type='checkbox'
        checked={fullPlaysOnly}
        onChange={e => setFullPlaysOnly(e.target.checked)}
      />
      <label htmlFor='full-plays-only'>Only include songs played in full</label>
      <canvas id='myChart' ref={ref} />
    </div>
  );
};

export default ChartComponent;
