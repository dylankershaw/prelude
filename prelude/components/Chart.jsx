import React, {useEffect, useRef, useState} from 'react';
import Chart from 'chart.js';

export default ({songData, artist}) => {
  const [chartData, setChartData] = useState([]);
  const ref = useRef(null);

  useEffect(() => {
    if (!artist) return setChartData([]);

    const songsWithArtist = songData.filter(s =>
      s.master_metadata_album_artist_name?.includes(artist)
    );

    const data = songsWithArtist.reduce((acc, cur) => {
      const date = cur.ts.split(' ')[0];
      return {...acc, [date]: (acc[date] || 0) + 1};
    }, {});

    setChartData(data);
  }, [songData, artist]);

  useEffect(() => {
    const myChartRef = ref.current.getContext('2d');

    new Chart(myChartRef, {
      type: 'line',
      data: {
        labels: Object.keys(chartData).reverse(),
        datasets: [
          {
            label: 'Songs Played',
            data: Object.values(chartData).reverse()
          }
        ]
      },
      options: {
        //Customize chart options
      }
    });
  }, [chartData]);

  return (
    <div>
      <canvas id='myChart' ref={ref} />
    </div>
  );
};
