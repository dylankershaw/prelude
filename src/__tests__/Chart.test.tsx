import React from 'react';
import {render} from '@testing-library/react';

import Chart from '../components/Chart';
import songData from './__mocks__/songData.json';

test('matches snapshot', () => {
  const {container} = render(<Chart selectedArtist='The Doors' songData={songData} />);
  expect(container).toMatchSnapshot();
});
