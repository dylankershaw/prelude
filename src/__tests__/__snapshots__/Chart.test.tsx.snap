// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <input
      id="full-plays-only"
      type="checkbox"
    />
    <label
      for="full-plays-only"
    >
      Only include songs played in full
    </label>
    <canvas
      class="chartjs-render-monitor"
      height="0"
      id="myChart"
      style="display: block;"
      width="0"
    />
    <style>
      
        .container {
          margin-top: 0.5rem;
        }

        canvas {
          margin-top: 2rem;
        }
      
    </style>
  </div>
</div>
`;
